### PR TITLE
Add krnlmon unit tests for DPDK target

### PR DESCRIPTION
**Pull request by mgaravin**

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -102,8 +102,22 @@ jobs:
              export SDE_INSTALL=${GITHUB_WORKSPACE}/install
              export DEPEND_INSTALL=${GITHUB_WORKSPACE}/deps
              source ./scripts/dpdk/setup_env.sh $IPDK_RECIPE $SDE_INSTALL $DEPEND_INSTALL
-             ./make-all.sh --target=dpdk  --prefix=${GITHUB_WORKSPACE}/dpdk-target 
-          
+             ./make-all.sh --target=dpdk  --prefix=${GITHUB_WORKSPACE}/dpdk-target
+
+      - name: 'Run Krnlmon Unit Test for DPDK Target'
```
Comment (ffoulkes) [inline .github/workflows/build_and_test.yml:107]:
```suggestion
      - name: 'Run Krnlmon Unit Tests for DPDK Target'
```

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -102,8 +102,22 @@ jobs:
...
+        run: |
+             cd ${GITHUB_WORKSPACE}/ipdk.recipe
+             cmake -B build -DDPDK_TARGET=on -DWITH_OVSP4RT=off
+             cmake --build build -j4 --target krnlmon-test
+             more build/Testing/Temporary/LastTest.log
```
Comment (ffoulkes) [inline .github/workflows/build_and_test.yml:112]:
```suggestion
             cat build/Testing/Temporary/LastTest.log
```
`more` is used in interactive environments. This is a batch environment.

**Review by ffoulkes — COMMENTED** [review]
I don't understand this. You've already done a complete build in the previous stage (`./make-all.sh target=dpdk`). The unit tests are compiled and ready to go. All you need to do now is `(cd build; ctest)`, as the [Running Unit Tests](https://wiki.ith.intel.com/pages/viewpage.action?pageId=2871826786#P4CPRunningUnitTests-Rununittests.1) wiki says.

The minimal build scenario is for cases where you _don't_ want to do a complete build.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (mgaravin) [reply .github/workflows/build_and_test.yml:107]: Hi Derek,
I have done exactly same way - pls. refer here - https://github.com/ipdk-io/networking-recipe/actions/runs/5419205766/workflow , but all tests failed!

**Comment by mgaravin** [comment]
> I don't understand this. You've already done a complete build in the previous stage (`./make-all.sh target=dpdk`). The unit tests are compiled and ready to go. All you need to do now is `(cd build; ctest)`, as the [Running Unit Tests](https://wiki.ith.intel.com/pages/viewpage.action?pageId=2871826786#P4CPRunningUnitTests-Rununittests.1) wiki says.
> 
> The minimal build scenario is for cases where you _don't_ want to do a complete build.


I have done exactly same way - pls. refer here - https://github.com/ipdk-io/networking-recipe/actions/runs/5419205766/workflow , but all tests failed!

**Comment by ffoulkes** [comment]
> I have done exactly same way - pls. refer here - https://github.com/ipdk-io/networking-recipe/actions/runs/5419205766/workflow , but all tests failed!

Can you send me a link to the log file? It would be useful to know what errors were reported.

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -34,7 +34,7 @@ jobs:
         working-directory: ./p4sde/tools/setup
         run: |
               sudo apt update -y
-              sudo apt install python python3-setuptools python3-pip python3-wheel python3-cffi sudo libssl-dev git zip libatomic1 libnl-route-3-dev curl tree -y
+              sudo apt install python python3-setuptools python3-pip python3-wheel python3-cffi sudo libssl-dev git zip libatomic1 libnl-route-3-dev curl tree libgtest-dev -y
```
Comment (ffoulkes) [inline .github/workflows/build_and_test.yml:37]: We supply Google Test (libgtest) as part of the Stratum dependencies (the same place we get gRPC, Protobuf, etc.). Adding it to the global environment only confuses tings.

**Comment by mgaravin** [comment]
> > I have done exactly same way - pls. refer here - https://github.com/ipdk-io/networking-recipe/actions/runs/5419205766/workflow , but all tests failed!
> 
> Can you send me a link to the log file? It would be useful to know what errors were reported.
> 
> I can see errors occurring if you're running the pipeline stages in parallel, but you appear to be sharing directories between stages, which would presumably lead to conflicts in both stages.

Looks like it works in independent build flow, but in shared workspace ( stratum and krnlmon) tests are failing!
Here is the test results from my fork - https://github.com/mgaravin/networking-recipe-fork/actions/runs/5474394680/jobs/9969115866

File: .github/workflows/build_and_test.yml (hunk 3 above)
Comment (mgaravin) [reply .github/workflows/build_and_test.yml:37]: I added this for testing, will remove in next patch!

**Review by ffoulkes — APPROVED** [review]

File: .github/workflows/build_and_test.yml (hunk 3 above)
Comment (mgaravin) [reply .github/workflows/build_and_test.yml]: removed!

File: .github/workflows/build_and_test.yml (hunk 2 above)
Comment (mgaravin) [reply .github/workflows/build_and_test.yml]: fixed!

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -102,13 +102,86 @@ jobs:
...
+          name: krnlmon_ut_logs
+          path: ipdk.recipe/krnlmon_ut_logs.zip
+
   py_bandit_test:
      timeout-minutes: 10
-     runs-on: ubuntu-20.04
+     runs-on: ubuntu-latest
      steps:
-       - name: 'checkout Networking Recipe'
+       - name: 'Checkout Networking Recipe'
```
Comment (ffoulkes) [inline .github/workflows/build_and_test.yml:184]:
```suggestion
       - name: 'Check Out Networking Recipe'
```
When used as a verb, this should be _check out_ (two words).
_checkout_ (one word) is a noun or adjective.

**Review by ffoulkes — APPROVED** [review]

File: .github/workflows/build_and_test.yml (hunk 4 above)
Comment (mgaravin) [reply .github/workflows/build_and_test.yml:184]: Also, Stratum tests failed with latest ubuntu, works only with ubuntu 20.04 - example: https://github.com/ipdk-io/networking-recipe/actions/runs/5527966086/jobs/10110937201

**Review by ffoulkes — CHANGES_REQUESTED** [review]

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -23,7 +23,7 @@ jobs:
     timeout-minutes: 100
     runs-on: ubuntu-20.04
     steps:
-      - name: 'Checkout p4-dpdk repo'
+      - name: 'Run Git CheckOut p4-dpdk repo'
```
Comment (ffoulkes) [inline .github/workflows/build_and_test.yml:26]: Why make it so complicated? "Check out p4-dpdk repo" is simpler and clearer.

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -102,13 +102,86 @@ jobs:
...
+      - name: 'Upload krnlmon-ut-result artifact'
+        uses: actions/upload-artifact@v3
+        with:
+          name: krnlmon_ut_logs
+          path: ipdk.recipe/krnlmon_ut_logs.zip
+
   py_bandit_test:
      timeout-minutes: 10
-     runs-on: ubuntu-20.04
+     runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (mgaravin) [reply .github/workflows/build_and_test.yml]: done!

File: .github/workflows/build_and_test.yml (hunk 5 above)
Comment (mgaravin) [reply .github/workflows/build_and_test.yml]: done!

**Review by nupuruttarwar — APPROVED** [review]
LGTM

**Comment by ffoulkes** [comment]
> @kkongusa and @ffoulkes, GitHub workflows are just wrapper to build and test instructions, running UT as part of development is not mandatory?

It's good developer practice, but no, it is not mandatory. It's also not particularly convenient, especially the Stratum unit tests. The logical thing to do is to use build automation to do this as part of the commit or merge process.

**Comment by ffoulkes** [comment]
@kkongusa In essence, workflow changes are self-testing. The updated workflow runs when you create the PR.

This won't detect issues that aren't covered by pre-merge checks. Is it possible to run GitHub actions on demand?

**Comment by ffoulkes** [comment]
> Is it possible to run GitHub actions on demand?

I just answered my own question: https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow

**Comment by kkongusa** [comment]
> @kkongusa In essence, workflow changes are self-testing. The updated workflow runs when you create the PR.
> 
> This won't detect issues that aren't covered by pre-merge checks. Is it possible to run GitHub actions on demand?

If we go by "production IS the testbed", just make sure when S!@T happens in production 1) there is some kind of "undo" button, and 2) someone who knows how to push it is actually available to push it. 8-)

**Pull request merged**